### PR TITLE
Add an RDS Postgres instance to Pulumi (for prod-aws)

### PR DESCRIPTION
**Pull request by wimax-grapl**
### Which issue does this PR correspond to?
Resolves https://github.com/grapl-security/issue-tracker/issues/813
Theoretically should resolve https://github.com/grapl-security/issue-tracker/issues/812 as well, but I've been wrong before!!!

### What changes does this PR make to Grapl? Why?
Add an RDS Postgres instance to Pulumi, and hook it up to plugin-registry.

### How were these changes tested?
I tested it on a new wimax-sandbox. plugin-registry successfully booted up with it and seemingly is talking to postgres!

### CODE REVIEW

File: pulumi/grapl/__main__.py [hunk 1]
```diff
@@ -181,16 +182,16 @@ def main() -> None:
...
             "rusoto_core=WARN",
             "rustls=WARN",
+            "serde_xml_rs=WARN",
         ]
     )
     py_log_level = "DEBUG"
 
-    # We've seen some potentially false failures from the default 5m timeout.
-    nomad_grapl_core_timeout = "8m"
+    nomad_grapl_core_timeout = "5m"
```
Comment (christophermaier) [inline pulumi/grapl/__main__.py:194]: Did we sort out what was causing this?

File: pulumi/infra/postgres.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        vpc_id: pulumi.Input[str],
+        subnet_ids: pulumi.Input[List[str]],
+        nomad_agent_security_group_id: pulumi.Input[str],
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ) -> None:
+        super().__init__("grapl:Postgres", name, None, opts)
+
+        # Burstable, 2GB memory, Gravitron 2.
+        # With 5GB of storage and all-day usage, it comes out to about $24/mo
+        instance_class = "db.t4g.small"
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:26]: Can we specify this as a stack configuration value instead?

File: pulumi/infra/postgres.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        rds_subnet_group = aws.rds.SubnetGroup(
+            "subnet-group",
+            subnet_ids=subnet_ids,
+            tags={"Name": f"{name}-subnet-group"},
+            opts=child_opts,
+        )
+
+        self.security_group = aws.ec2.SecurityGroup(
+            f"postgres-security-group",
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:48]: `f"{name}-security-group"`

(Or maybe `f"{name}-postgres-security-group"`, if we're adding that as a `Name` tag below? If so, other resource names should probably include `postgres` as well)

File: pulumi/infra/postgres.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+        # Burstable, 2GB memory, Gravitron 2.
+        # With 5GB of storage and all-day usage, it comes out to about $24/mo
+        instance_class = "db.t4g.small"
+
+        child_opts = pulumi.ResourceOptions(parent=self)
+
+        self.username = "postgres"
+        self.password = random.RandomPassword(
+            "password",
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:32]: `f"{name}-password"`

File: pulumi/infra/postgres.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            length=32,
+            # Disable special characters, ":" can lead to unexpected
+            # sqlx errors: https://github.com/launchbadge/sqlx/issues/1624
+            special=False,
+            opts=child_opts,
+        ).result
+
+        rds_subnet_group = aws.rds.SubnetGroup(
+            "subnet-group",
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:41]: `f"{name}-subnet-group"`

File: pulumi/infra/postgres.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            # Tags are necessary for the moment so we can look up the resource from a different pulumi stack.
+            # Once this is refactored we can remove the tags
+            tags={"Name": f"{name}-postgres-security-group"},
+            opts=pulumi.ResourceOptions(parent=self),
+        )
+        postgres_port = 5432
+
+        # Allow communication between nomad-agents and RDS
+        aws.ec2.SecurityGroupRule(
+            "nomad-agents-egress-to-rds",
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:59]: `{name}` ought to be part of all these resource names.

File: pulumi/infra/postgres.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            security_group_id=self.security_group.id,
+            from_port=postgres_port,
+            to_port=postgres_port,
+            protocol="tcp",
+            source_security_group_id=nomad_agent_security_group_id,
+            opts=pulumi.ResourceOptions(parent=self.security_group),
+        )
+
+        # There's some funkiness where, if not specified, you may end up with
+        # RDS trying to stand up an instance in an AZ not included in `subnets`.
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:81]: Is there an issue we can reference for this?

File: pulumi/infra/postgres.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        def to_az(ids: List[str]) -> pulumi.Output[str]:
+            subnet_id = ids[0]
+            subnet = aws.ec2.Subnet.get("subnet", subnet_id)
+            # for some reason mypy gets hung up on the typing of this
+            az: pulumi.Output[str] = subnet.availability_zone
+            return az
+
+        availability_zone: pulumi.Output[str] = pulumi.Output.from_input(
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:89]: Would it be easier to just pass an availability zone into the resource constructor directly, rather than going through this?

File: pulumi/infra/postgres.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            subnet_ids
+        ).apply(to_az)
+
+        self.instance = aws.rds.Instance(
+            "instance",
+            name=name,  # only alphanumeric
+            engine="postgres",
+            engine_version="13.4",
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:97]: This might be good to pull up into stack configuration, too, as well as the storage parameters.

File: pulumi/infra/postgres.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            vpc_security_group_ids=[self.security_group.id],
+            # TODO: EnableIAMDatabaseAuthentication
+            # Makes teardown easier. This can be revisited.
+            skip_final_snapshot=True,
+            # Eventually this would be managed by Vault.
+            # In the mean time, security is basically only enforced by VPC
+            username=self.username,
+            password=self.password,
+            port=postgres_port,
+            # Note: See Pulumi docs about `apply_immediately` flag for ASAP migrations
```
Comment (christophermaier) [inline pulumi/infra/postgres.py:113]: What is this pertaining to?

File: pulumi/grapl/__main__.py [hunk 11]
```diff
@@ -338,6 +340,13 @@ def main() -> None:
             nomad_agent_security_group_id=nomad_agent_security_group_id,
         )
 
+        postgres = Postgres(
+            name="postgres",
```
Comment (christophermaier) [inline pulumi/grapl/__main__.py:344]: Are we only going to have one Postgres instance for all of Grapl, or one per service? If the latter, we should name this `plugin-registry`.

File: pulumi/grapl/__main__.py (hunk 1 above)
Comment (wimax-grapl) [reply pulumi/grapl/__main__.py:194]: _probably_ it was Plugin Registry never successfully coming up, because it didn't have a real postgres instance to connect to - see 812

Now that artifacts work we'll be able to further debug this stuff if it happens again in the future.

In the meantime, this does work on my sandbox.

File: pulumi/infra/postgres.py (hunk 2 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:26]: Sure.

File: pulumi/infra/postgres.py (hunk 4 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:32]: done

File: pulumi/infra/postgres.py (hunk 5 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:41]: done

File: pulumi/infra/postgres.py (hunk 7 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:81]: It's not really an "issue" per se. Funkiness is the wrong word. Let me rephrase it.

File: pulumi/infra/postgres.py (hunk 6 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:59]: done. for some reason I have this misconception that "oh if it's nested as a child opts, we don't need to shoehorn in name" but that's wrong

File: pulumi/infra/postgres.py (hunk 10 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:113]: Removing this, the comment just introduces confusion and we don't need to worry about this for the time being

File: pulumi/infra/postgres.py (hunk 9 above)
Comment (wimax-grapl) [reply pulumi/infra/postgres.py:97]: I pulled engine_version out.
The storage stuff needs some genuine consideration - there's a bajillion other parameters - so I'm going to leave it hardcoded with a warning about it.

File: pulumi/grapl/__main__.py (hunk 11 above)
Comment (wimax-grapl) [reply pulumi/grapl/__main__.py:344]: Open question. I asked Colin. I'm going to submit the rest of the updated PR without addressing this [yet]

File: pulumi/bin/copy_artifacts_from_rc.sh [hunk 12]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env bash
 # This will grab a subset of keys from `origin/rc`'s Pulumi.testing.yaml.
 set -euo pipefail
+set -o xtrace
```
Comment (wimax-grapl) [inline pulumi/bin/copy_artifacts_from_rc.sh:4]: This script has proven to be somewhat brittle, so I'd actually like to leave this option on for future users, if that's alright!

File: pulumi/infra/config.py [hunk 13]
```diff
@@ -213,6 +213,10 @@ def require_artifact(artifact_name: str) -> Any:
     return artifact
 
 
+def get_postgres_instance_class() -> str:
```
Comment (christophermaier) [inline pulumi/infra/config.py]: Looks like other similar functions don't have the `get_` prefix; consistency would be good.

Also, should we make these all properties? 🤔 (EDIT: oh, these are all just bare functions; I was thinking we had an object. Refactoring for another time, I suppose.)

File: pulumi/grapl/README.md [hunk 14]
```diff
@@ -44,6 +44,16 @@ according to your anticipated workload.
 pulumi config set postgres-instance-type db.t4g.small
 ```
 
+### `postgres-version`
+
+Not required for local-grapl.
+
+Which version of Postgres to use. Must be >= 13.4.
```
Comment (christophermaier) [inline pulumi/grapl/README.md:51]: Is this version constraint asserted anywhere in the code?

File: pulumi/bin/copy_artifacts_from_rc.sh [hunk 15]
```diff
@@ -49,15 +50,15 @@ main() {
     echo "--- Grab artifacts from origin/rc config file"
     git fetch origin
     git show origin/rc:pulumi/grapl/Pulumi.testing.yaml > ${RC_CONFIG_FILE}
-    local -r artifacts=$(pulumi config --config-file="${RC_CONFIG_FILE}" get artifacts)
+    local -r artifacts=$(pulumi config --config-file="${RC_CONFIG_FILE}" --stack="${CURRENT_STACK}" get artifacts)
```
Comment (christophermaier) [inline pulumi/bin/copy_artifacts_from_rc.sh:53]: Yup, always specify a stack.

**Review by christophermaier — APPROVED** [review]

File: pulumi/infra/config.py (hunk 13 above)
Comment (wimax-grapl) [reply pulumi/infra/config.py]: this was actually removed and replaced fwiw

Comment (wimax-grapl) [reply pulumi/infra/config.py]:
```
@dataclass
class PostgresConfigValues:
    instance_type: str
    postgres_version: str

    @staticmethod
    def from_config() -> PostgresConfigValues:
        return PostgresConfigValues(
            instance_type=pulumi.Config().require("postgres-instance-type"),
            postgres_version=pulumi.Config().require("postgres-version"),
        )
```

File: pulumi/grapl/README.md (hunk 14 above)
Comment (wimax-grapl) [reply pulumi/grapl/README.md:51]: It's not. Now that the hard stuff is done I can add a quick major-minor-comparator or something
